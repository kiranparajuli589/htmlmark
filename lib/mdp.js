import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"

/**
 * Markdown Parser for Javascript Application
 */
export class MDP {
	indent
	codeHighlightFn
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
	}
	/**
	 * Runs the HTML parsing for the given lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	h(lines) {
		const lexer = new Lexer(lines)
		const parser = new Parser(lexer.run(), null, this.codeHighlightFn)
		return parser.run()
	}

	/**
	 * Runs HTML parsing for the lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	hT(lines) {
		const start = Date.now()
		const html = this.h(lines)
		const end = Date.now()
		console.info(`\x1b[35mParsing time: ${end - start}ms\x1b[0m`)
		return html
	}

	/**
	 * For demo preview
	 *
	 * @param {String} str - The string to parse
	 * @param {Function|null} codeHighlightFn - A function to highlight code blocks
	 *
	 * @returns {{ elapsedTime: number, html: string, lex: Object}}
	 */
	hP(str, codeHighlightFn = null) {
		const lines = str.split("\n")
		const start = Date.now()
		const lexer = new Lexer(lines)
		const lex = lexer.run()
		const parser = new Parser(lex, null, codeHighlightFn)
		const html = parser.run()
		const end = Date.now()
		return {
			elapsedTime: end - start,
			html,
			lex
		}
	}
}
