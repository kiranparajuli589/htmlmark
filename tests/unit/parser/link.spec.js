import { MDP } from "../../mdp.js"


const mdp = new MDP()


describe("link", () => {
	it("should parse a link", () => {
		const lines = [
			"[Kiran Parajuli](https://kiranparajuli.com.np)"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse a link with text", () => {
		const lines = [
			"[Kiran Parajuli](https://kiranparajuli.com.np) is a good"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse a link with title", () => {
		const lines = [
			"test **bold *and* bold** with *italics* and a [Kiran Parajuli](https://kiranparajuli.com.np 'K Pictures')",
			"",
			"here *is a **beautiful** piece of* art"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
})
