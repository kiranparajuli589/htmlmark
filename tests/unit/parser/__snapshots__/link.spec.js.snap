// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link should parse a link 1`] = `"<p><a href=\\"https://kiranparajuli.com.np\\">Kiran Parajuli</a></p>"`;

exports[`link should parse a link with text 1`] = `"<p><a href=\\"https://kiranparajuli.com.np\\">Kiran Parajuli</a> is a good</p>"`;

exports[`link should parse a link with title 1`] = `"<p>test <strong>bold <em>and</em> bold</strong> with <em>italics</em> and a <a href=\\"https://kiranparajuli.com.np\\" title=\\"K Pictures\\">Kiran Parajuli</a></p><p>here <em>is a <strong>beautiful</strong> piece of</em> art</p>"`;
